src/runtime/pybuffer.cs

```diff
@@ -0,0 +1,228 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Python.Runtime
+{
+    public class PyBuffer : IDisposable
+    {
+        private PyObject _exporter;
+        private IntPtr _handle;
+        private Runtime.Py_buffer _view;
+        private GCHandle _gchandle;
+
+        internal PyBuffer(PyObject exporter, int flags)
+        {
+            int size = Marshal.SizeOf(typeof(Runtime.Py_buffer));
+            byte[] rawData = new byte[size];
+            _gchandle = GCHandle.Alloc(rawData, GCHandleType.Pinned);
+            _handle = _gchandle.AddrOfPinnedObject();
+
+            if (Runtime.PyObject_GetBuffer(exporter.Handle, _handle, flags) < 0)
+            {
+                _gchandle.Free();
+                throw new PythonException();
+            }
+
+            _exporter = exporter;
+            _view = (Runtime.Py_buffer)Marshal.PtrToStructure(_handle, typeof(Runtime.Py_buffer));
+
+            if (_view.shape != IntPtr.Zero)
+            {
+                Shape = new long[_view.ndim];
+                Marshal.Copy(_view.shape, Shape, 0, (int)_view.len * sizeof(long));
```

Comment:
This is the same issue as noted below, on 32bit Windows (and actually also on 32bit Linux) this will fail, it might actually segfault.